### PR TITLE
Add Plugs, catch exception while consuming and make handlers mandatory

### PR DESCRIPTION
**Pull request by gpopides**
### TLDR: 

Bugfixes:
- Catch exceptions when consuming messages

Removals/Changes:
- `onSuccess` and `onFailure` handlers for producers and consumers are now mandatory during the creation and won't be part of configuration using the configuration lambda
- Removed SQS Pipeline

Features:
- Added Plugs
- Added a bean to stop tabour when spring boot stops running
- Added new convinience functions to proto package to map sqs messages

## Plugs
Plugs are a new `plugin` system that allows users to execute custom methods after a message is `consumed` or `produced` successfully or there is a failure during the process

Users need to implement the following interfaces
- ProducerPlug
- ConsumerPlug

The implementations are added to each consumer/producer individually using `plugs.add(myplug)` during the configuration of the consumer/producer.

Currently, there are no built-in plugs

## New proto functions
Proto package has new functions to handle deserialization of sqs messages. 

The functions allow the user to handle failures in different manners. Depending on the function, they can deserialize and if an exception will be raised they can handle it with the following manners:
- Catch the exception themselves
- Return a null
- Return a null and use an error handler to handle the error

A new method is also added to `map` the deserialized object to a new object

### CODE REVIEW

**Review by Bogdanfl935 — APPROVED** [review]

File: plug/src/main/kotlin/com/katanox/tabour/plug/Plug.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package com.katanox.tabour.plug
+
+sealed class PlugRecord<T> {
```
[the remark this answers is not in the document]
Comment (gpopides) [reply plug/src/main/kotlin/com/katanox/tabour/plug/Plug.kt:3]: One thing i can think of is the slack notification. Using. a plug on the confirmation consumer to send the created booking notificationss

File: plug/src/main/kotlin/com/katanox/tabour/plug/Plug.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    abstract val key: T
+}
+
+data class SuccessPlugRecord<T, K>(val data: T, override val key: K) : PlugRecord<K>()
```
[the remark this answers is not in the document]
Comment (gpopides) [reply plug/src/main/kotlin/com/katanox/tabour/plug/Plug.kt:7]: No, plugs work independently and based on the message consumed/produced and not the output of another plug.
